[Feature] Turns a gitfs remote into a true "formula", merges specified branch into current saltenv

ekristen opened the pull request:
Thank you to @terminalmage for taking the time to give me basically everything I needed to implement this in a comment on the original issue. Very much appreciated.

**Note:** (added after @terminalmage initial review comment) -- we originally discussed naming the variable `global` but that seems to be reserved and causes python syntax errors when being referenced on `self` (ie `return self.global`). I used `formula` for the initial PR, but I agree it's not the best name as this feature has many more uses. 

### What does this PR do?
It enables a gitfs remote to be a true formula, or a global one, when set, all states from the referenced branch are merged into whichever saltenv is actually being referenced by the minion in which the states are being resolved for.

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/40527

### How to use and test this feature.
Add the following to your salt-minion (for a masterless test) config.

```yaml
environment: testing
fileserver_backend:
  - git
gitfs_remotes:
  - https://github.com/ekristen/salt-states-testing.git
  - https://github.com/ekristen/salt-formula-testing.git
```

If you inspect `https://github.com/ekristen/salt-states-testing.git` you'll find a testing branch and if you inspect `https://github.com/ekristen/salt-formula-testing.git` you will NOT find a testing branch. Given this setup, if you run the following command `salt-call --local state.sls testing` it will succeed (as it is from the first repo), however if you run `salt-call --local state.sls formula` it will fail as there is no `testing` branch on the second repository.

Modify the config so it looks like this.

```yaml
environment: testing
fileserver_backend:
  - git
gitfs_remotes:
  - https://github.com/ekristen/salt-states-testing.git
  - https://github.com/ekristen/salt-formula-testing.git:
    - formula: master
```

Then rerun the commands, first `salt-call --local state.sls testing` it will still resolve and run, and now `salt-call --local state.sls formula` this time it will resolve and run appropriately. 

### Tests written?
No - I looked at the gitfs tests and there did not seem to be an appropriate way or place to test this new feature within the current test suite for gitfs.

ekristen committed:
adds ability to turn remote into true formula, merges remote into all saltenvs if enabled

salt-jenkins commented:
@ekristen, thanks for your PR! By analyzing the history of the files in this pull request, we identified @terminalmage, @rallytime and @thatch45 to be potential reviewers.

terminalmage commented on salt/fileserver/gitfs.py:
```diff
@@ -50,7 +50,7 @@
 PER_REMOTE_OVERRIDES = ('base', 'mountpoint', 'root', 'ssl_verify',
                         'env_whitelist', 'env_blacklist', 'refspecs')
-PER_REMOTE_ONLY = ('name', 'saltenv')
+PER_REMOTE_ONLY = ('formula', 'name', 'saltenv')
```
I thought we discussed naming this as ``global``, not ``formula``. This feature is useful for more than just formulas, so giving it such a specific name seems wrong.

ekristen replied:
@terminalmage my bad! I forgot to mention when I typed up the original PR, I tried global, but it kept throwing a syntax error, seems to be a reserved keyword.

ekristen replied:
Perhaps a synonym? `world`, `wordwide`, or `universal` (I think I like universal the best?)

terminalmage replied:
Oh, interesting. I've never tried using ``global`` as a class attribute, so I've never run into this syntax error before. I'll try to brainstorm some ideas as well.

terminalmage replied:
What about ``all_saltenvs``?

ekristen replied:
That works for me, the value would still be the branch name I assume?

terminalmage replied:
Yeah. If you're comfortable writing RST, this should be documented in the gitfs documentation page at doc/topics/tutorials/gitfs.rst. If not, I can take care of documentation.

ekristen replied:
I'm fine with writing RST. I can take a stab at adding some initial docs in.

terminalmage replied:
Cool. I can help polish them if they need to be tweaked in any way.

Thanks for taking this on, I'm really excited about this feature!

ekristen replied:
Glad to help, this will be very useful for me.

terminalmage commented:
Don't worry about tests, we have limited test coverage for gitfs but I've been working on a framework which will make it easier to do integration tests on git_pillar and gitfs (https://github.com/saltstack/salt/pull/40777). Writing new integration tests for gitfs is something I have plans for soon anyway, and I can make sure something like this is covered.

ekristen committed:
rename to all_saltenvs, add initial documentation

ekristen commented:
@terminalmage I did notice the original issue was slated for `Oxygen` but I assume that was based on if you had to do all the work, with an open PR now, might it be possible to get it merged for an upcoming release?

For now I left all documentation notes to reference Oxygen.

terminalmage commented:
We've passed the feature freeze for Nitrogen, so develop is the correct branch and Oxygen is the correct release.

ekristen commented:
Is there a timeline? https://docs.saltstack.com/en/latest/topics/releases/version_numbers.html has TBD for both Nitrogen and Oxygen. Thanks!

terminalmage commented:
Nitrogen is expected late this month, with Oxygen in the fall (tentatively).

cachedout commented:
@terminalmage Can you re-review this and approve if you think it's ready?

terminalmage committed:
No need to assign an ignored exception to a variable

terminalmage committed:
Clarify that it is only all_saltenvs which is new in Oxygen

Depending on how you read it, the previous version may have suggested
that all of the per-remote parameters are new in Oxygen. This makes it
more explicit that the all_saltenvs param is the only one new to Oxygen.

terminalmage committed:
mention all_saltenvs and link back to gitfs walkthrough

terminalmage committed:
Enhance the Global Remotes docs

terminalmage committed:
Add debug logging for cases where all_saltenvs overrides the per-saltenv configuration

terminalmage commented:
@ekristen I opened https://github.com/ekristen/salt/pull/1 against your fork which tweaks and builds upon the docs you added, and also modifies the gitfs code to add debug logging when a per-saltenv configuration is being overridden by ``all_saltenvs``. This is something I had not considered initially until I started testing the functionality of your PR.

Can you please merge that PR and then test and confirm that it still works to your liking? If so, and assuming tests pass, I think this is good to merge.

Thanks once again for your contribution.

ekristen committed:
Merge pull request #1 from terminalmage/pr-40985

Improve all_saltenvs docs

ekristen commented:
@terminalmage I have merged your PR, I will test shortly and confirm all is working appropriately still.

ekristen commented:
@terminalmage everything works, the only comment I have is that the debug messages off a bit?

They seem to repeat multiple times, and always state `None` was mapped to `saltenv`. As you are way more familiar with the system I'll defer to you as to whether or not this is correct, just wanted to mention it.

```
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'base' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'base' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'base' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'base' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'base' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'testing' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'testing' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
```

terminalmage commented:
That definitely is not correct. Change this line:

```python
if all_saltenvs_ref != per_saltenv_ref:
```

to

```python
if per_saltenv_ref and all_saltenvs_ref != per_saltenv_ref:
```

I didn't test it without a per-saltenv config, duh.  Thanks for catching this.

ekristen committed:
only log debug if per-saltenv config is defined

cachedout commented:
@terminalmage Looks like the requested change has been made. Requesting re-review.

terminalmage reviewed (approved):
LGTM. Thanks, @ekristen!

ekristen commented:
Thanks @terminalmage
